File: src/ocr/image-preprocess.js
// Image preprocessing for text recognition — contrast enhancement, binarization
// Used to improve handwriting and camera photo quality before PP-OCRv5 / Tesseract

// Convert image to grayscale ImageData
function toGrayscale(imageData) {
  const { data, width, height } = imageData;
  const gray = new Uint8ClampedArray(width * height);
  for (let i = 0; i < width * height; i++) {
    const p = i * 4;
    gray[i] = Math.round(data[p] * 0.299 + data[p + 1] * 0.587 + data[p + 2] * 0.114);
  }
  return { gray, width, height };
}

// Otsu's method — find optimal threshold for binarization
function otsuThreshold(gray, width, height) {
  const histogram = new Int32Array(256);
  const n = width * height;
  for (let i = 0; i < n; i++) histogram[gray[i]]++;

  let sum = 0;
  for (let i = 0; i < 256; i++) sum += i * histogram[i];

  let wB = 0, sumB = 0;
  let maxVariance = 0, threshold = 128;

  for (let t = 0; t < 256; t++) {
    wB += histogram[t];
    if (wB === 0) continue;
    const wF = n - wB;
    if (wF === 0) break;
    sumB += t * histogram[t];
    const mB = sumB / wB;
    const mF = (sum - sumB) / wF;
    const variance = wB * wF * (mB - mF) * (mB - mF);
    if (variance > maxVariance) {
      maxVariance = variance;
      threshold = t;
    }
  }
  return threshold;
}

// Apply binary threshold to ImageData
function applyThreshold(imageData, threshold) {
  const { data } = imageData;
  const n = Math.floor(data.length / 4);
  for (let i = 0; i < n; i++) {
    const p = i * 4;
    const v = data[p] * 0.299 + data[p + 1] * 0.587 + data[p + 2] * 0.114;
    const bw = v >= threshold ? 255 : 0;
    data[p] = bw;
    data[p + 1] = bw;
    data[p + 2] = bw;
  }
  return imageData;
}

// Contrast-Limited Adaptive Histogram Equalization (simplified)
// Enhances local contrast in tiles, reduces noise in flat regions
function claheGray(gray, width, height, tileSize, clipLimit) {
  tileSize = tileSize || 32;
  clipLimit = clipLimit || 2.0;

  const result = new Uint8ClampedArray(width * height);
  const tilesX = Math.ceil(width / tileSize);
  const tilesY = Math.ceil(height / tileSize);

  // Build CDF for each tile
  const tileCDFs = [];
  for (let ty = 0; ty < tilesY; ty++) {
    for (let tx = 0; tx < tilesX; tx++) {
      const hist = new Int32Array(256);
      const x0 = tx * tileSize;
      const y0 = ty * tileSize;
      const x1 = Math.min(x0 + tileSize, width);
      const y1 = Math.min(y0 + tileSize, height);
      let count = 0;

      for (let y = y0; y < y1; y++) {
        for (let x = x0; x < x1; x++) {
          hist[gray[y * width + x]]++;
          count++;
        }
      }

      // Clip histogram
      if (clipLimit > 0 && count > 0) {
        const clipThreshold = Math.floor(clipLimit * count / 256);
        let excess = 0;
        for (let i = 0; i < 256; i++) {
          if (hist[i] > clipThreshold) {
            excess += hist[i] - clipThreshold;
            hist[i] = clipThreshold;
          }
        }
        const redist = Math.floor(excess / 256);
        for (let i = 0; i < 256; i++) {
          hist[i] += redist;
        }
      }

      // Build CDF
      const cdf = new Float32Array(256);
      cdf[0] = hist[0] / count;
      for (let i = 1; i < 256; i++) {
        cdf[i] = cdf[i - 1] + hist[i] / count;
      }
      tileCDFs.push({ cdf, x0, y0, x1, y1, cx: (x0 + x1) / 2, cy: (y0 + y1) / 2 });
    }
  }

  // Apply bilinear interpolation between tile CDFs
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      // Find 4 nearest tile centers
      const tx = Math.floor(x / tileSize);
      const ty = Math.floor(y / tileSize);
      const neighbors = [];

      for (let dy = -1; dy <= 1; dy++) {
        for (let dx = -1; dx <= 1; dx++) {
          const nx = tx + dx;
          const ny = ty + dy;
          if (nx >= 0 && nx < tilesX && ny >= 0 && ny < tilesY) {
            const tile = tileCDFs[ny * tilesX + nx];
            const dist = Math.sqrt((x - tile.cx) ** 2 + (y - tile.cy) ** 2) || 1;
            neighbors.push({ cdf: tile.cdf, dist: 1 / dist });
          }
        }
      }

      const v = gray[y * width + x];
      let sum = 0, weightSum = 0;
      for (const n of neighbors) {
        sum += n.cdf[v] * n.dist;
        weightSum += n.dist;
      }
      result[y * width + x] = Math.round((sum / weightSum) * 255);
    }
  }
  return result;
}

// Write grayscale array back to ImageData (preserves alpha channel)
function grayToImageData(gray, imageData) {
  const { data, width, height } = imageData;
  for (let i = 0; i < width * height; i++) {
    const p = i * 4;
    data[p] = gray[i];
    data[p + 1] = gray[i];
    data[p + 2] = gray[i];
  }
  return imageData;
}

// ── Public API ──

// Enhance contrast for handwriting images (faint strokes → crisp dark text)
// Returns a new canvas with enhanced image
export function enhanceHandwriting(sourceImg) {
  const w = sourceImg.naturalWidth || sourceImg.width;
  const h = sourceImg.naturalHeight || sourceImg.height;

  const canvas = document.createElement('canvas');
  canvas.width = w;
  canvas.height = h;
  const ctx = canvas.getContext('2d');
  ctx.drawImage(sourceImg, 0, 0);

  const imageData = ctx.getImageData(0, 0, w, h);
  const { gray } = toGrayscale(imageData);

  // Apply CLAHE with small tiles and moderate clipping
  const enhanced = claheGray(gray, w, h, 16, 3.0);
  grayToImageData(enhanced, imageData);
  ctx.putImageData(imageData, 0, 0);

  return canvas;
}

// Binarize image (Otsu thresholding) — returns pure black & white canvas
export function binarize(sourceImg) {
  const w = sourceImg.naturalWidth || sourceImg.width;
  const h = sourceImg.naturalHeight || sourceImg.height;

  const canvas = document.createElement('canvas');
  canvas.width = w;
  canvas.height = h;
  const ctx = canvas.getContext('2d');
  ctx.drawImage(sourceImg, 0, 0);

  const imageData = ctx.getImageData(0, 0, w, h);
  const { gray } = toGrayscale(imageData);
  const threshold = otsuThreshold(gray, w, h);
  applyThreshold(imageData, threshold);
  ctx.putImageData(imageData, 0, 0);

  return canvas;
}

// Preprocess camera photo for text recognition
// 1. Convert to grayscale  2. Enhance contrast (CLAHE)  3. Binarize
export function preprocessForOCR(sourceImg) {
  const w = sourceImg.naturalWidth || sourceImg.width;
  const h = sourceImg.naturalHeight || sourceImg.height;

  const canvas = document.createElement('canvas');
  canvas.width = w;
  canvas.height = h;
  const ctx = canvas.getContext('2d');
  ctx.drawImage(sourceImg, 0, 0);

  const imageData = ctx.getImageData(0, 0, w, h);
  const { gray } = toGrayscale(imageData);

  // Step 1: CLAHE contrast enhancement
  const enhanced = claheGray(gray, w, h, 32, 2.5);

  // Step 2: Otsu binarization
  const threshold = otsuThreshold(enhanced, w, h);

  // Apply binary threshold
  for (let i = 0; i < w * h; i++) {
    const p = i * 4;
    const bw = enhanced[i] >= threshold ? 255 : 0;
    imageData.data[p] = bw;
    imageData.data[p + 1] = bw;
    imageData.data[p + 2] = bw;
  }

  ctx.putImageData(imageData, 0, 0);
  return canvas;
}

// Check if image looks like handwriting (thin strokes on white bg)
export function looksLikeHandwriting(sourceImg) {
  const w = sourceImg.naturalWidth || sourceImg.width;
  const h = sourceImg.naturalHeight || sourceImg.height;

  const canvas = document.createElement('canvas');
  const size = Math.min(128, w, h);
  canvas.width = size;
  canvas.height = size;
  const ctx = canvas.getContext('2d');
  ctx.drawImage(sourceImg, 0, 0, size, size);

  const imageData = ctx.getImageData(0, 0, size, size);
  const { data } = imageData;
  const n = size * size;

  // Count dark pixels and average darkness
  let darkCount = 0;
  let totalIntensity = 0;
  for (let i = 0; i < n; i++) {
    const p = i * 4;
    const v = data[p] * 0.299 + data[p + 1] * 0.587 + data[p + 2] * 0.114;
    totalIntensity += v;
    if (v < 128) darkCount++;
  }

  const avgIntensity = totalIntensity / n;
  const darkRatio = darkCount / n;

  // Handwriting: mostly light background, very few dark pixels, average intensity is high
  return avgIntensity > 200 && darkRatio > 0.002 && darkRatio < 0.15;
}

File: android/app/src/main/java/com/latexsnipper/app/ocr/ImagePreProcess.java
package com.latexsnipper.app.ocr;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;

/**
 * ImagePreProcess — contrast enhancement, binarization, handwriting detection.
 * <p>
 * Matches image-preprocess.js: enhanceHandwriting() + binarize() + preprocessForOCR() + looksLikeHandwriting().
 * <p>
 * Uses CLAHE (Contrast-Limited Adaptive Histogram Equalization) and Otsu thresholding.
 */
public class ImagePreProcess {

    /**
     * Check if image looks like handwriting.
     * Matches looksLikeHandwriting(): light bg, very few dark pixels, avg intensity high.
     */
    public static boolean looksLikeHandwriting(Bitmap bitmap) {
        int size = Math.min(128, Math.min(bitmap.getWidth(), bitmap.getHeight()));
        Bitmap thumb = Bitmap.createScaledBitmap(bitmap, size, size, true);

        int[] argb = new int[size * size];
        thumb.getPixels(argb, 0, size, 0, 0, size, size);
        if (thumb != bitmap) thumb.recycle();

        int n = size * size;
        int darkCount = 0;
        float totalIntensity = 0;

        for (int i = 0; i < n; i++) {
            int pixel = argb[i];
            float r = (pixel >> 16) & 0xFF;
            float g = (pixel >> 8) & 0xFF;
            float b = pixel & 0xFF;
            float v = r * 0.299f + g * 0.587f + b * 0.114f;
            totalIntensity += v;
            if (v < 128) darkCount++;
        }

        float avgIntensity = totalIntensity / n;
        float darkRatio = (float) darkCount / n;

        return avgIntensity > 200 && darkRatio > 0.002f && darkRatio < 0.15f;
    }

    /**
     * Enhance contrast for handwriting using CLAHE.
     * Matches enhanceHandwriting().
     */
    public static Bitmap enhanceHandwriting(Bitmap source) {
        int w = source.getWidth();
        int h = source.getHeight();
        int[] argb = new int[w * h];
        source.getPixels(argb, 0, w, 0, 0, w, h);

        float[] gray = toGrayFloats(argb);
        float[] enhanced = clahe(gray, w, h, 16, 3.0f);

        return buildBitmapFromGray(enhanced, w, h);
    }

    /**
     * Binarize using Otsu thresholding. Matches binarize().
     */
    public static Bitmap binarize(Bitmap source) {
        int w = source.getWidth();
        int h = source.getHeight();
        int[] argb = new int[w * h];
        source.getPixels(argb, 0, w, 0, 0, w, h);

        float[] gray = toGrayFloats(argb);
        int threshold = otsuThreshold(gray);
        return buildBitmapFromThreshold(gray, threshold, w, h);
    }

    /**
     * Full OCR preprocessing: CLAHE + Otsu. Matches preprocessForOCR().
     */
    public static Bitmap preprocessForOCR(Bitmap source) {
        int w = source.getWidth();
        int h = source.getHeight();
        int[] argb = new int[w * h];
        source.getPixels(argb, 0, w, 0, 0, w, h);

        float[] gray = toGrayFloats(argb);
        float[] enhanced = clahe(gray, w, h, 32, 2.5f);
        int threshold = otsuThreshold(enhanced);

        return buildBitmapFromThreshold(enhanced, threshold, w, h);
    }

    // ── Grayscale conversion ──

    private static float[] toGrayFloats(int[] argb) {
        float[] gray = new float[argb.length];
        for (int i = 0; i < argb.length; i++) {
            int p = argb[i];
            gray[i] = ((p >> 16) & 0xFF) * 0.299f
                    + ((p >> 8)  & 0xFF) * 0.587f
                    + (p         & 0xFF) * 0.114f;
        }
        return gray;
    }

    // ── Otsu threshold ──

    private static int otsuThreshold(float[] gray) {
        int[] hist = new int[256];
        int n = gray.length;
        for (float v : gray) hist[Math.round(v)]++;

        float sum = 0;
        for (int i = 0; i < 256; i++) sum += i * hist[i];

        float wB = 0, sumB = 0;
        float maxVariance = 0;
        int threshold = 128;

        for (int t = 0; t < 256; t++) {
            wB += hist[t];
            if (wB == 0) continue;
            float wF = n - wB;
            if (wF == 0) break;
            sumB += t * hist[t];
            float mB = sumB / wB;
            float mF = (sum - sumB) / wF;
            float variance = wB * wF * (mB - mF) * (mB - mF);
            if (variance > maxVariance) {
                maxVariance = variance;
                threshold = t;
            }
        }
        return threshold;
    }

    // ── CLAHE ──

    private static float[] clahe(float[] gray, int width, int height,
                                  int tileSize, float clipLimit) {
        float[] result = new float[gray.length];
        int tilesX = (int) Math.ceil((float) width / tileSize);
        int tilesY = (int) Math.ceil((float) height / tileSize);
        int totalTiles = tilesX * tilesY;

        // Store per-tile CDFs
        float[][] tileCDFs = new float[totalTiles][256];
        float[] tileCX = new float[totalTiles];
        float[] tileCY = new float[totalTiles];

        for (int ty = 0; ty < tilesY; ty++) {
            for (int tx = 0; tx < tilesX; tx++) {
                int tileIdx = ty * tilesX + tx;
                int x0 = tx * tileSize;
                int y0 = ty * tileSize;
                int x1 = Math.min(x0 + tileSize, width);
                int y1 = Math.min(y0 + tileSize, height);

                int[] hist = new int[256];
                int count = 0;
                for (int y = y0; y < y1; y++) {
                    for (int x = x0; x < x1; x++) {
                        int val = Math.round(gray[y * width + x]);
                        hist[val]++;
                        count++;
                    }
                }

                // Clip histogram
                if (clipLimit > 0 && count > 0) {
                    int clipThreshold = (int) (clipLimit * count / 256);
                    int excess = 0;
                    for (int i = 0; i < 256; i++) {
                        if (hist[i] > clipThreshold) {
                            excess += hist[i] - clipThreshold;
                            hist[i] = clipThreshold;
                        }
                    }
                    int redist = excess / 256;
                    for (int i = 0; i < 256; i++) hist[i] += redist;
                }

                // Build CDF
                float[] cdf = tileCDFs[tileIdx];
                if (count > 0) {
                    cdf[0] = (float) hist[0] / count;
                    for (int i = 1; i < 256; i++) {
                        cdf[i] = cdf[i - 1] + (float) hist[i] / count;
                    }
                }

                tileCX[tileIdx] = (x0 + x1) / 2.0f;
                tileCY[tileIdx] = (y0 + y1) / 2.0f;
            }
        }

        // Apply bilinear interpolation
        for (int y = 0; y < height; y++) {
            for (int x = 0; x < width; x++) {
                int tx0 = Math.max(0, Math.min(tilesX - 1, x / tileSize));
                int ty0 = Math.max(0, Math.min(tilesY - 1, y / tileSize));

                float sum = 0, weightSum = 0;
                for (int dy = -1; dy <= 1; dy++) {
                    for (int dx = -1; dx <= 1; dx++) {
                        int nx = tx0 + dx;
                        int ny = ty0 + dy;
                        if (nx >= 0 && nx < tilesX && ny >= 0 && ny < tilesY) {
                            int idx = ny * tilesX + nx;
                            float dist = (float) Math.sqrt(
                                (x - tileCX[idx]) * (x - tileCX[idx]) +
                                (y - tileCY[idx]) * (y - tileCY[idx]));
                            if (dist < 1) dist = 1;
                            float w = 1.0f / dist;
                            int val = Math.round(gray[y * width + x]);
                            if (val < 0) val = 0;
                            if (val > 255) val = 255;
                            sum += tileCDFs[idx][val] * w;
                            weightSum += w;
                        }
                    }
                }
                float v = (sum / weightSum) * 255;
                result[y * width + x] = Math.min(255, Math.max(0, v));
            }
        }

        return result;
    }

    // ── Bitmap construction helpers ──

    private static Bitmap buildBitmapFromGray(float[] gray, int w, int h) {
        int[] argb = new int[gray.length];
        for (int i = 0; i < gray.length; i++) {
            int v = Math.round(Math.min(255, Math.max(0, gray[i])));
            argb[i] = Color.rgb(v, v, v);
        }
        Bitmap b = Bitmap.createBitmap(w, h, Bitmap.Config.ARGB_8888);
        b.setPixels(argb, 0, w, 0, 0, w, h);
        return b;
    }

    private static Bitmap buildBitmapFromThreshold(float[] gray, int threshold, int w, int h) {
        int[] argb = new int[gray.length];
        for (int i = 0; i < gray.length; i++) {
            int v = gray[i] >= threshold ? 255 : 0;
            argb[i] = Color.rgb(v, v, v);
        }
        Bitmap b = Bitmap.createBitmap(w, h, Bitmap.Config.ARGB_8888);
        b.setPixels(argb, 0, w, 0, 0, w, h);
        return b;
    }
}
